library/core/src/str/mod.rs

```diff
@@ -76,6 +76,14 @@ use iter::MatchIndicesInternal;
 use iter::SplitInternal;
 use iter::{MatchesInternal, SplitNInternal};
 
+/// Hello I am a str
+#[cfg(not(bootstrap))]
+#[unstable(feature = "the_str_type_itself_in_all_its_glory", issue = "none")]
+#[rustc_has_incoherent_inherent_impls]
+#[lang = "str"]
+#[repr(transparent)]
+pub struct Str([u8]);
```

Comment:
is there a reason that the struct is `Str` (uppercase) here instead of the usual `str` used in all other places?

Reply:
well currently it's an implementation detail, so it doesn't really matter what it's named -- the unstable feature gate means you can't even name it via its path

Reply:
Are `Str` and `str` intended to be the same or different types?


(And the unstable feature gate doesn't necessarily apply in std)

Reply:
> Are Str and str intended to be the same or different types?

They're the same type. All name resolution for `str` currently happens through a special path that handles other primitive types (like `i32`). If/when we wanted to change `str` to be resolved via the prelude or something, then yeah, the name would probably have to be changed (though maybe not, since we could just `use core::str::Str as str;` in the prelude).

Similarly for printing, `str` as a lang item is specialized to always print as that, to make sure we avoid printing `core::str::Str` in code when [path trimming](https://blog.aloni.org/posts/path-trimming-in-rust-nightly/) is explicitly disabled, as it is in certain diagnostics.

Reply:
Thanks for explaining, that makes sense. I suspect we'd want it to be `str` everywhere eventually (even though it doesn't matter) just to keep things clearer in rustdoc source and so on.

(But it's understandable that it might take some time to get to that point)